README.md
# Evm Bn [![GitHub Actions][gha-badge]][gha] [![Coverage][codecov-badge]][codecov] [![Styled with Prettier][prettier-badge]][prettier] [![License: MIT][license-badge]][license]

[gha]: https://github.com/PaulRBerg/evm-bn/actions
[gha-badge]: https://github.com/PaulRBerg/evm-bn/actions/workflows/ci.yml/badge.svg
[codecov]: https://codecov.io/gh/PaulRBerg/evm-bn
[codecov-badge]: https://codecov.io/gh/PaulRBerg/evm-bn/branch/main/graph/badge.svg
[prettier]: https://prettier.io
[prettier-badge]: https://img.shields.io/badge/Code_Style-Prettier-ff69b4.svg
[license]: https://opensource.org/licenses/MIT
[license-badge]: https://img.shields.io/badge/License-MIT-blue.svg

Evm Bn is a utility for converting between stringified fixed-point numbers and
[Ethers.js](https://github.com/ethers-io/ethers.js) BigNumbers, as well as the reverse process. It is particularly
useful for projects based on the Ethereum Virtual Machine (EVM), given that 1 ETH is equivalent to 1e18 wei.

- Accepts scientific notation.
- Limits the precision to 78 digits.
- Enforces 60 integer digits and 18 fractional digits.
- Designed to be used alongside
  [@ethersproject/bignumber](https://github.com/ethers-io/ethers.js/tree/master/packages/bignumber).
- Slices the fractional digits automatically at position `n + 1` and above, with `n` the number of decimals, rounding
  down in the process.

## Install

```sh
$ pnpm add evm-bn
```

## Usage

### To Bn

```ts
import type { BigNumber } from "@ethersproject/bignumber";
import { toBn } from "evm-bn";

// 3141500000000000000
const foo: BigNumber = toBn("3.1415");

// 115792089237316195423570985008687907853269984665640564039457584007913129639935
const bar: BigNumber = toBn("115792089237316195423570985008687907853269984665640564039457.584007913129639935");

// 100000000000000
const baz: BigNumber = toBn("100e6", 6);
```

### From Bn

```ts
import type { BigNumber } from "@ethersproject/bignumber";
import { fromBn } from "evm-bn";

// 3.1415
const foo: BigNumber = fromBn(BigNumber.from("3141500000000000000"));

// 115792089237316195423570985008687907853269984665640564039457.584007913129639935
const bar: BigNumber = fromBn(
  BigNumber.from("115792089237316195423570985008687907853269984665640564039457584007913129639935"),
);

// 100000000
const baz: BigNumber = fromBn(BigNumber.from("100000000000000"), 6);
```

## Contributing

Feel free to dive in! [Open](https://github.com/PaulRBerg/evm-bn/issues/new) an issue,
[start](https://github.com/PaulRBerg/evm-bn/discussions/new) a discussion or submit a PR.

### Set Up

Clone the repositories and install the dependencies:

```sh
$ pnpm install
```

Now you can start making changes.

## License

This project is licensed under MIT.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Common Changelog](https://common-changelog.org/), and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

[1.2.0]: https://github.com/PaulRBerg/evm-bn/compare/v1.1.2...v1.2.0
[1.1.2]: https://github.com/PaulRBerg/evm-bn/compare/v1.1.1...v1.1.2
[1.1.1]: https://github.com/PaulRBerg/evm-bn/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/PaulRBerg/evm-bn/compare/v1.0.1...v1.1.0
[1.0.1]: https://github.com/PaulRBerg/evm-bn/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/PaulRBerg/evm-bn/releases/tag/v1.0.0

## [1.2.0] - 2023-03-28

### Changed

- Bump dependencies (@PaulRBerg)
- Improve documentation and formatting (@PaulRBerg)
- Switch from Yarn to Pnpm for dependency management (@PaulRBerg)

### Added

- Support uppercase `E` in exponential notation ([#16](https://github.com/PaulRBerg/evm-bn/pull/16)) (@tiagox)

## [1.1.2] - 2022-07-26

### Changed

- Adhere to Common Changelog in `CHANGELOG.md` (@PaulRBerg)
- Change the license from "Unlicense" to "MIT" (@PaulRBerg)
- Fix number of decimals in error message in `fromBn` function (@PaulRBerg)
- Polish the README (@PaulRBerg)

## [1.1.1] - 2021-10-27

### Changed

- Semver from "^5.5.0" to "5.x" for `@ethersproject/bignumber` peer dependency (@PaulRBerg)

## [1.1.0] - 2021-10-26

### Added

- TypeScript declaration maps and source maps (@PaulRBerg)

### Changed

- Upgrade to `@ethersproject/bignumber` v5.5.0 (@PaulRBerg)

## [1.0.1] - 2021-09-29

### Fixed

- Relax type checking in `fromBn` function (@PaulRBerg)

## [1.0.0] - 2021-09-27

### Added

- First release of the package (@PaulRBerg)


test/toBn.test.ts
import { expect } from "earljs";
import forEach from "mocha-each";

import { toBn } from "../src";

describe("toBn", function () {
  describe("when x is not a string", function () {
    const testSets = [undefined, null, true, 2.71, Math.PI, { x: 100 }, function () {}];

    forEach(testSets).it("throws an error", function (x: any) {
      expect(() => toBn(x)).toThrow("Input must be a string");
    });
  });

  describe("when x is a string", function () {
    describe("when x is not a base 10 number", function () {
      const testSets = ["qwerty", "Bonjour le Monde", "a0b0z0d0", "0xcafe", "áèâéèê"];

      forEach(testSets).it("throws an error", function (x: string) {
        expect(() => toBn(x)).toThrow("Unknown format for fixed-point number: " + x);
      });
    });

    describe("when x is a base 10 number", function () {
      describe("when the number of decimals is out of bounds", function () {
        const testSets = [-1729, 0, 78, 1729];

        forEach(testSets).it("throws an error", function (decimals: number) {
          const x: string = "3.141592653589793238";
          expect(() => toBn(x, decimals)).toThrow("Decimals must be between 1 and 77");
        });
      });

      describe("when the number of decimals is within the bounds", function () {
        describe("when the number of decimals is 1", function () {
          const decimals: number = 1;
          const testSets = [
            ["1", "10"],
            ["2.8", "28"],
            ["3.1", "31"],
            [
              "11579208923731619542357098500868790785326998466564056403945758400791312963993.5",
              "115792089237316195423570985008687907853269984665640564039457584007913129639935",
            ],
          ];

          forEach(testSets).it("takes %f and returns %f", function (x: string, expected: string) {
            expect(expected).toEqual(String(toBn(x, decimals)));
          });
        });

        describe("when the number of decimals is 77", function () {
          const decimals: number = 77;
          const testSets = [
            ["1", "100000000000000000000000000000000000000000000000000000000000000000000000000000"],
            ["2.8", "280000000000000000000000000000000000000000000000000000000000000000000000000000"],
            ["3.1", "310000000000000000000000000000000000000000000000000000000000000000000000000000"],
            [
              "1.15792089237316195423570985008687907853269984665640564039457584007913129639935",
              "115792089237316195423570985008687907853269984665640564039457584007913129639935",
            ],
          ];

          forEach(testSets).it("takes %f and returns %f", function (x: string, expected: string) {
            expect(expected).toEqual(String(toBn(x, decimals)));
          });
        });

        describe("when the number of decimals is the default, which is 18", function () {
          describe("when x has more than 60 integer digits", function () {
            const testSets: string[] = [
              "1e60",
              "3.14e78",
              "3860393555824785230250785487392862718135031063639377380058239.75498055625639",
              "1641859000171773711896013123523298990385936318426437089678010964091794.294085164534340909",
              "6579166703548843398264284621832829277269325500191914304280053707339887173863112402178486913.1782903",
            ];

            forEach(testSets).it("takes %f and throws an error", function (x: string) {
              expect(() => toBn(x)).toThrow("Unknown format for fixed-point number: " + x);
            });
          });

          describe("when x does not have more than 60 integer digits", function () {
            describe("when x has more than 18 fractional digits", function () {
              describe("when x is an exponential", function () {
                const testSets = [
                  ["1e-19", "0"],
                  ["3.14e-78", "0"],
                  ["412.1239812412889000291381209e6", "412123981241288900029138120"],
                ];

                forEach(testSets).it("takes %f and returns %f", function (x: string, expected: string) {
                  expect(expected).toEqual(String(toBn(x)));
                });
              });

              describe("when x is a real number", function () {
                const testSets = [
                  ["0.4902990866551706545", "490299086655170654"],
                  ["8.00019706910353772932315427126187897", "8000197069103537729"],
                  [
                    "91.7496885414232378075050290903596022687487995387586369206142513715480845755362175",
                    "91749688541423237807",
                  ],
                ];

                forEach(testSets).it("takes %f and returns %f", function (x: string, expected: string) {
                  expect(expected).toEqual(String(toBn(x)));
                });
              });
            });

            describe("when x does not have more than 18 fractional digits", function () {
              describe("when x is an exponential", function () {
                const testSets = [
                  ["8.92e-5", "89200000000000"],
                  ["8.92E-5", "89200000000000"],
                  ["1e2", "100000000000000000000"],
                  ["1E2", "100000000000000000000"],
                  ["3.14e5", "314000000000000000000000"],
                  ["3.14E5", "314000000000000000000000"],
                  ["89.12e12", "89120000000000000000000000000000"],
                  ["89.12E12", "89120000000000000000000000000000"],
                  ["100.05e18", "100050000000000000000000000000000000000"],
                  ["100.05E18", "100050000000000000000000000000000000000"],
                  ["2e3", "2000000000000000000000"],
                  ["2E3", "2000000000000000000000"],
                  ["42733.02954e45", "42733029540000000000000000000000000000000000000000000000000000000000"],
                  ["42733.02954E45", "42733029540000000000000000000000000000000000000000000000000000000000"],
                  [
                    "1.15792089237316195423570985008687907853269984665640564039457584007913129639935e59",
                    "115792089237316195423570985008687907853269984665640564039457584007913129639935",
                  ],
                  [
                    "1.15792089237316195423570985008687907853269984665640564039457584007913129639935E59",
                    "115792089237316195423570985008687907853269984665640564039457584007913129639935",
                  ],
                ];

                forEach(testSets).it("takes %f and returns %f", function (x: string, expected: string) {
                  expect(expected).toEqual(String(toBn(x)));
                });
              });

              describe("when x is a real number", function () {
                const testSets = [
                  ["0.0000892", "89200000000000"],
                  ["1", "1000000000000000000"],
                  ["2.8931", "2893100000000000000"],
                  ["3.141592653589793238", "3141592653589793238"],
                  ["781.0042190122", "781004219012200000000"],
                  ["2000", "2000000000000000000000"],
                  ["26241.06698", "26241066980000000000000"],
                  [
                    "115792089237316195423570985008687907853269984665640564039457.584007913129639935",
                    "115792089237316195423570985008687907853269984665640564039457584007913129639935",
                  ],
                ];

                forEach(testSets).it("takes %f and returns %f", function (x: string, expected: string) {
                  expect(expected).toEqual(String(toBn(x)));
                });
              });
            });
          });
        });
      });
    });
  });
});


test/fromBn.test.ts
import { BigNumber } from "@ethersproject/bignumber";
import { expect } from "earljs";
import forEach from "mocha-each";

import { fromBn } from "../src";

describe("fromBn", function () {
  describe("when x is undefined", function () {
    it("throws an error", function () {
      const x: any = undefined;
      expect(() => fromBn(x)).toThrow("Input must not be undefined");
    });
  });

  describe("when x is not undefined", function () {
    describe("when the number of decimals is out of bounds", function () {
      const testSets = [-1729, 0, 78, 1729];

      forEach(testSets).it("throws an error", function (decimals: number) {
        const x = BigNumber.from("3141592653589793238");
        expect(() => fromBn(x, decimals)).toThrow("Decimals must be between 1 and 77");
      });
    });

    describe("when the number of decimals is within the bounds", function () {
      describe("when the number of decimals is 1", function () {
        const decimals: number = 1;
        const testSets = [
          ["10", "1"],
          ["28", "2.8"],
          ["31", "3.1"],
          [
            "115792089237316195423570985008687907853269984665640564039457584007913129639935",
            "11579208923731619542357098500868790785326998466564056403945758400791312963993.5",
          ],
        ];

        forEach(testSets).it("takes %f and returns %f", function (x: string, expected: string) {
          expect(expected).toEqual(fromBn(BigNumber.from(x), decimals));
        });
      });

      describe("when the number of decimals is 77", function () {
        const decimals: number = 77;
        const testSets = [
          ["100000000000000000000000000000000000000000000000000000000000000000000000000000", "1"],
          ["280000000000000000000000000000000000000000000000000000000000000000000000000000", "2.8"],
          ["310000000000000000000000000000000000000000000000000000000000000000000000000000", "3.1"],
          [
            "115792089237316195423570985008687907853269984665640564039457584007913129639935",
            "1.15792089237316195423570985008687907853269984665640564039457584007913129639935",
          ],
        ];

        forEach(testSets).it("takes %f and returns %f", function (x: string, expected: string) {
          expect(expected).toEqual(fromBn(BigNumber.from(x), decimals));
        });
      });
    });
  });
});


src/toBn.ts
import type { BigNumber } from "@ethersproject/bignumber";
import { parseFixed } from "@ethersproject/bignumber";
import fromExponential from "from-exponential";

/**
 * Convert a stringified fixed-point number to a big number with a custom number of decimals.
 *
 * @remarks
 * - Accepts scientific notation.
 * - Checks are in place to adhere to the numerical constraints of the EVM.
 */
export function toBn(x: string, decimals: number = 18): BigNumber {
  if (x === undefined || typeof x !== "string") {
    throw new Error("Input must be a string");
  }

  if (decimals < 1 || decimals > 77) {
    throw new Error("Decimals must be between 1 and 77");
  }

  let xs: string = x;

  // Convert from exponential notation.
  if (x.includes("e") || x.includes("E")) {
    xs = fromExponential(x);
  }

  // Limit the number of decimals to the value provided.
  if (xs.includes(".")) {
    const parts: string[] = xs.split(".");
    parts[1] = parts[1].slice(0, decimals);
    xs = parts[0] + "." + parts[1];
  }

  // Check if x is a whole number or a fixed-point number with some maximum number of decimals.
  const digits: number = 78 - decimals;
  const regexp: RegExp = new RegExp(`^[-+]?(\\d{1,${digits}}|(?=\\d+\\.\\d+)\\d{1,${digits}}\\.\\d{1,${decimals}})$`);

  if (regexp.test(xs)) {
    return parseFixed(xs, decimals);
  } else {
    throw new Error("Unknown format for fixed-point number: " + x);
  }
}
